euphoria-flink/src/main/java/cz/seznam/euphoria/flink/streaming/windowing/WindowOperator.java

```diff
@@ -0,0 +1,465 @@
+
+        processTriggerResult(currentWindow, triggerResult, mergingWindowSet);
+      }
+      mergingWindowSet.persist();
+
+    } else {
+      for (WID window : element.getWindows()) {
+        setupEnvironment(getCurrentKey(), window);
+
+        if (localMode) {
+          // register cleanup timer for each window
+          endOfStreamTimerService.registerEventTimeTimer(window, Long.MAX_VALUE);
+        }
+
+        getWindowState(window).add(element.getValue());
+
+        // process trigger
+        Trigger.TriggerResult triggerResult = trigger.onElement(
+                element.getTimestamp(),
+                window,
+                triggerContext);
+
+        processTriggerResult(window, triggerResult, null);
+      }
+    }
+  }
+
+  @Override
+  public void onEventTime(InternalTimer<KEY, WID> timer) throws Exception {
+    WID window = timer.getNamespace();
+    Object key = timer.getKey();
+
+    setupEnvironment(key, window);
+
+    Trigger.TriggerResult triggerResult;
+    if (timer.getTimestamp() == Long.MAX_VALUE) {
+      // at the end of time flush window immediately
+      triggerResult = Trigger.TriggerResult.FLUSH_AND_PURGE;
+    } else {
+      triggerResult = trigger.onTimer(
+              timer.getTimestamp(),
+              window,
+              triggerContext);
+    }
+
+    MergingWindowSet<WID> mergingWindowSet = null;
+    if (windowing instanceof MergingWindowing) {
+      mergingWindowSet = getMergingWindowSet();
+    }
+
+    processTriggerResult(window, triggerResult, mergingWindowSet);
+
+    if (mergingWindowSet != null) {
+      mergingWindowSet.persist();
+    }
+  }
+
+  @Override
+  public void onProcessingTime(InternalTimer<KEY, WID> timer) throws Exception {
+    throw new UnsupportedOperationException("We are not using processing time at all");
+  }
+
+  @Override
+  public void processWatermark(Watermark mark) throws Exception {
+    // FIXME relies heavily on Flink internal implementation
+    // which may change in the future and break this
+
+    // we need to advance watermark of timer services in given order
+    // timerService first to fire all scheduled triggers
+    ((HeapInternalTimerService) timerService).advanceWatermark(mark.getTimestamp());
+    if (localMode) {
+      ((HeapInternalTimerService) endOfStreamTimerService).advanceWatermark(mark.getTimestamp());
+    }
+
+    output.emitWatermark(mark);
+  }
+
+  private void processTriggerResult(WID window,
+                                    Trigger.TriggerResult tr,
+                                    @Nullable MergingWindowSet<WID> mergingWindowSet) {
+    WID stateWindow = window;
+    State windowState;
+
+    if (windowing instanceof MergingWindowing) {
+      Objects.requireNonNull(mergingWindowSet);
+      stateWindow = mergingWindowSet.getStateWindow(window);
+      windowState = getWindowState(stateWindow);
+    } else {
+      windowState = getWindowState(window);
+    }
```

Comment:
potential micro-optimization possibility here: don't lookup out the window state from the state table when not necessary (e.g. `tr == NOOP`)